src/Compilers/Core/Portable/Operations/Operation.cs

```diff
@@ -46,41 +76,203 @@ public Operation(OperationKind kind, SyntaxNode syntax, ITypeSymbol type, Option
+                base(OperationKind.None, semanticMode, node, type: null, constantValue: constantValue)
+            {
+                _getChildren = getChildren;
+            }
 
-            public void Accept(OperationVisitor visitor)
+            public override void Accept(OperationVisitor visitor)
             {
                 visitor.VisitNoneOperation(this);
             }
 
-            public TResult Accept<TArgument, TResult>(OperationVisitor<TArgument, TResult> visitor, TArgument argument)
+            public override TResult Accept<TArgument, TResult>(OperationVisitor<TArgument, TResult> visitor, TArgument argument)
             {
                 return visitor.VisitNoneOperation(this, argument);
             }
 
-            public IEnumerable<IOperation> Children => _getChildren().NullToEmpty();
+            public override IEnumerable<IOperation> Children
+            {
+                get
+                {
+                    foreach (var child in _getChildren().NullToEmpty().WhereNotNull())
+                    {
+                        yield return Operation.SetParentOperation(child, this);
+                    }
+                }
+            }
+        }
+
+        private static readonly ObjectPool<Queue<IOperation>> s_queuePool =
+            new ObjectPool<Queue<IOperation>>(() => new Queue<IOperation>(), 10);
+
+        private IOperation WalkDownOperationToFindParent(HashSet<IOperation> operationAlreadyProcessed, IOperation root)
+        {
+            void EnqueueChildOperations(Queue<IOperation> queue, IOperation parent)
+            {
+                // children can return null
+                foreach (var o in parent.Children.WhereNotNull())
+                {
+                    queue.Enqueue(o);
+                }
+            }
+
+            var operationQueue = s_queuePool.Allocate();
+
+            try
+            {
+                EnqueueChildOperations(operationQueue, root);
+
+                // walk down the tree to find parent operation
+                // every operation returned by the queue should already have Parent operation set
+                while (operationQueue.Count > 0)
+                {
+                    var operation = operationQueue.Dequeue();
+
+                    if (!operationAlreadyProcessed.Add(operation))
+                    {
+                        // don't process IOperation we already processed otherwise,
+                        // we can walk down same tree multiple times
+                        continue;
+                    }
+
+                    if (operation == this)
+                    {
+                        // parent found
+                        return operation.Parent;
+                    }
+
+                    // It can't filter visiting children by node span since IOperation
+                    // might have children which belong to completely different sub tree of
+                    // syntax tree
```

Comment:
hrmm... one of our invariants is that if you have an IOperatoin 'o' that corresponds to a SyntaxNode 'n' then all child operations of 'o' would point to child syntax nodes of 'n'.  So you should never have any children which belong to different sub trees.